packages/react/src/core/ContextMenu.tsx

```diff
@@ -0,0 +1,132 @@
+import { useRef, useState, type ComponentProps } from 'react';
+import { Root, Portal, Content, Item, Trigger } from '@radix-ui/react-context-menu';
+import type { I18n } from '@tutorialkit/types';
+
+interface FileChangeEvent {
+  type: 'FILE' | 'FOLDER';
+  method: 'ADD' | 'REMOVE' | 'RENAME';
+  value: string;
+}
+
+interface FileRenameEvent extends FileChangeEvent {
+  method: 'RENAME';
+  oldValue: string;
+}
+
+interface Props extends ComponentProps<'div'> {
+  /** Callback invoked when file is changed. */
+  onFileChange?: (event: FileChangeEvent | FileRenameEvent) => void;
+
+  /** Directory of the clicked file. */
+  directory: string;
+
+  /** Whether to render new files/directories before or after the trigger element. Defaults to `'before'`. */
+  position?: 'before' | 'after';
+
+  /** Localized texts for menu. */
+  i18n?: Pick<I18n, 'fileTreeCreateFileText' | 'fileTreeCreateFolderText'>;
+
+  /** Props for trigger wrapper. */
+  triggerProps?: ComponentProps<'div'> & { 'data-testid'?: string };
+}
+
+export function ContextMenu({
+  onFileChange,
+  directory,
+  i18n,
+  position = 'before',
+  children,
+  triggerProps,
+  ...props
+}: Props) {
+  const [state, setState] = useState<'IDLE' | 'ADD_FILE' | 'ADD_FOLDER'>('IDLE');
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  if (!onFileChange) {
+    return children;
+  }
+
+  function onFileNameEnd(event: React.KeyboardEvent<HTMLInputElement> | React.FocusEvent<HTMLInputElement>) {
+    const name = event.currentTarget.value;
+    const isFile = state === 'ADD_FILE';
+
+    // files must contain extension
```

Comment:
I don't think we should not make that assumption. Extensions are not mandatory anywhere in tutorialkit AFAIK.

Reply:
I agree - we should allow files without extensions. Need https://github.com/stackblitz/tutorialkit/pull/314#discussion_r1753809729 for that. 👷